### ISSUE TITLE
[MNG-5965] Parallel build multiplies work if multiple goals are given

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#7896]
**[Matthias Schmalz](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=matthias.schmalz@sap.com)** opened **[MNG-5965](https://issues.apache.org/jira/browse/MNG-5965?redirect=false)** and commented

When I run a parallel build which invokes multiple goals Maven multiplies the work e.g. when I run

install sonar:sonar
Every phase is executed once as expected. However when I run

install sonar:sonar -T 4
Every phase is executed twice.

The problem can be reproduced with a single simple Java project (of course parallel builds are useless with a single module). Debugging showed that every Mojo is really called twice per project.
Find attached a simple project and two build outputs, where you can see that the parallel build is doing everything twice (you can ignore that Sonar fails in the end, due to no server is up).

---

**Affects:** 3.3.9

**Attachments:**
- [parallel.test.zip](https://issues.apache.org/jira/secure/attachment/12783845/parallel.test.zip) (_16.05 kB_)

**Issue Links:**
- [MNG-5705](https://issues.apache.org/jira/browse/MNG-5705) NPE on parallel build in BuilderCommon.handleBuildError(BuilderCommon.java:147)

- [MNG-5705](https://issues.apache.org/jira/browse/MNG-5705) NPE on parallel build in BuilderCommon.handleBuildError(BuilderCommon.java:147)
 (_**"is depended upon by"**_)

**Remote Links:**
- [GitHub Pull Request #22
](https://github.com/apache/maven-integration-testing/pull/22)
- [GitHub Pull Request #125
](https://github.com/apache/maven/pull/125)
- [GitHub Pull Request #22
](https://github.com/apache/maven-integration-testing/pull/22)
- [GitHub Pull Request #125
](https://github.com/apache/maven/pull/125)

3 votes, 8 watchers

**Comment by jira-importer** [comment]
**[Janos Gyerik](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=janos.gyerik)** commented

Although `sonar:sonar` is mentioned here, it's not required for the reproducer, and it has nothing to do with the issue here. You can reproduce for example with `mvn clean install:install-file -Dfile=path/to/jar -T10`, and get an output as below. Notice the duplicate steps.

```
[INFO] Scanning for projects...
[INFO] 
[INFO] Using the MultiThreadedBuilder implementation with a thread count of 10
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] Building Example Project 1.0-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO] 
[INFO] --- maven-clean-plugin:3.0.0:clean (default-clean) @ my-project ---
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] Building Example Project 1.0-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO] 
[INFO] --- maven-clean-plugin:3.0.0:clean (default-clean) @ my-project ---
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] Building Example Project 1.0-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO] 
[INFO] --- maven-install-plugin:2.5.2:install-file (default-cli) @ my-project ---
[INFO] Installing /path/to/jar to /path/to/maven/repo
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] Building Example Project 1.0-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO] 
[INFO] --- maven-install-plugin:2.5.2:install-file (default-cli) @ my-project ---
[INFO] Installing /path/to/jar to /path/to/maven/repo
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 0.638 s (Wall Clock)
[INFO] Finished at: 2017-06-27T09:55:48+02:00
[INFO] Final Memory: 11M/240M
[INFO] ------------------------------------------------------------------------
```

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Fixed with [c6bdafed23f152d83d766c1ed3db8c1c4b723eb3](https://gitbox.apache.org/repos/asf?p=maven.git;a=commit;h=c6bdafed23f152d83d766c1ed3db8c1c4b723eb3) and [d3d6ea3a6345a51a08e8dd41a0254e8cb0f2692f](https://gitbox.apache.org/repos/asf?p=maven-integration-testing.git;a=commit;h=d3d6ea3a6345a51a08e8dd41a0254e8cb0f2692f)..
